File: src/parser_utils/indexed_ast_mapper.ml
(*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

class index_mapper =
  object
    inherit [Loc.t, Loc.t, ILoc.t, ILoc.t] Flow_polymorphic_ast_mapper.mapper

    val mutable counter = 0

    method on_loc_annot loc =
      let c = counter in
      counter <- counter + 1;
      (loc, c)

    method on_type_annot loc =
      let c = counter in
      counter <- counter + 1;
      (loc, c)
  end

class unindex_mapper =
  object
    inherit [ILoc.t, ILoc.t, Loc.t, Loc.t] Flow_polymorphic_ast_mapper.mapper

    method on_loc_annot (loc, _id) = loc

    method on_type_annot (loc, _id) = loc
  end

File: rust_port/crates/flow_parser_utils/src/indexed_ast_mapper.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use dupe::Dupe;
use flow_parser::loc::Loc;
use flow_parser::polymorphic_ast_mapper::LocMapper;

use crate::iloc::ILoc;

pub struct IndexMapper {
    counter: u32,
}

impl IndexMapper {
    pub fn new() -> Self {
        Self { counter: 0 }
    }
}

impl Default for IndexMapper {
    fn default() -> Self {
        Self::new()
    }
}

impl LocMapper<Loc, Loc, ILoc, ILoc> for IndexMapper {
    fn on_loc_annot(&mut self, loc: &Loc) -> Result<ILoc, !> {
        let c = self.counter;
        self.counter += 1;
        Ok(ILoc(loc.dupe(), c))
    }

    fn on_type_annot(&mut self, loc: &Loc) -> Result<ILoc, !> {
        let c = self.counter;
        self.counter += 1;
        Ok(ILoc(loc.dupe(), c))
    }
}

pub struct UnindexMapper;

impl LocMapper<ILoc, ILoc, Loc, Loc> for UnindexMapper {
    fn on_loc_annot(&mut self, iloc: &ILoc) -> Result<Loc, !> {
        let ILoc(loc, _id) = iloc;
        Ok(loc.dupe())
    }

    fn on_type_annot(&mut self, iloc: &ILoc) -> Result<Loc, !> {
        let ILoc(loc, _id) = iloc;
        Ok(loc.dupe())
    }
}
